fms/utils/generation.py

```diff
@@ -126,3 +128,234 @@ def truncate_after_eos(result, eos_token_id):
         eos_idx = eos_idx[0].item()
         result = result[: eos_idx + 1]
     return result
+
+
+def speculative_generate(
```

Comment:
seems very long, maybe could be broken up with some helper functions to make it easier to read (and could document sub-parts). e.g. maybe functions to get the mask, to get the position IDs, one for the loop that manipulates the kv cache.